### PR TITLE
Fix FD_SETSIZE too many open files bug (#15500)

### PR DESCRIPTION
**Pull request by skoch-twocream**
See https://github.com/pimcore/pimcore/issues/15500

### CODE REVIEW

**Comment by robertSt7** [comment]
Hi @skochTC if we pass an opened stream and not a string to this function I think it's not the right place to close it here. May you can check at the top of the function what is passed and then check before closing. What do you think?

**Comment by skoch-twocream** [comment]
> Hi @skochTC if we pass an opened stream and not a string to this function I think it's not the right place to close it here. May you can check at the top of the function what is passed and then check before closing. What do you think?

Hi, I am not sure. 

What do you think by doing that where the method is called?
```patch
diff --git a/lib/Helper/TemporaryFileHelperTrait.php b/lib/Helper/TemporaryFileHelperTrait.php
index ed7f4a2695..218af35998 100644
--- a/lib/Helper/TemporaryFileHelperTrait.php
+++ b/lib/Helper/TemporaryFileHelperTrait.php
@@ -39,7 +39,6 @@ trait TemporaryFileHelperTrait
 
         if (is_resource($stream)) {
             $streamMeta = stream_get_meta_data($stream);
-            fclose($stream);
             $stream = $streamMeta['uri'];
         }
 
diff --git a/models/Asset/Image/Thumbnail/Processor.php b/models/Asset/Image/Thumbnail/Processor.php
index a9a30fa1a3..6ddd0f907e 100644
--- a/models/Asset/Image/Thumbnail/Processor.php
+++ b/models/Asset/Image/Thumbnail/Processor.php
@@ -268,7 +268,9 @@ class Processor
                 }
 
                 if (is_resource($fileSystemPath)) {
+                    $fileSystemPathStream = $fileSystemPath;
                     $fileSystemPath = self::getLocalFileFromStream($fileSystemPath);
+                    fclose($fileSystemPathStream);
                 }
 
                 if (!file_exists($fileSystemPath)) {
diff --git a/models/Asset/Thumbnail/ImageThumbnailTrait.php b/models/Asset/Thumbnail/ImageThumbnailTrait.php
index b90dafa9bd..237420e255 100644
--- a/models/Asset/Thumbnail/ImageThumbnailTrait.php
+++ b/models/Asset/Thumbnail/ImageThumbnailTrait.php
@@ -403,7 +403,10 @@ trait ImageThumbnailTrait
             return null;
         }
 
-        return self::getLocalFileFromStream($stream);
+        $localFile = self::getLocalFileFromStream($stream);
+        fclose($stream);
+
+        return $localFile;
     }
 
     /**
diff --git a/models/User.php b/models/User.php
index cc2c6b40e0..817c7d5943 100644
--- a/models/User.php
+++ b/models/User.php
@@ -566,7 +566,9 @@ final class User extends User\UserRole
         $storage = Tool\Storage::get('admin');
         if ($storage->fileExists($this->getOriginalImageStoragePath())) {
             if (!$storage->fileExists($this->getThumbnailImageStoragePath())) {
-                $localFile = self::getLocalFileFromStream($storage->readStream($this->getOriginalImageStoragePath()));
+                $originalImageStream = $storage->readStream($this->getOriginalImageStoragePath());
+                $localFile = self::getLocalFileFromStream($originalImageStream);
+                fclose($originalImageStream);
                 $targetFile = File::getLocalTempFilePath('png');
 
                 $image = \Pimcore\Image::getInstance();
```

There are two places where I'm not sure exactly how the stream is processed:
https://github.com/pimcore/pimcore/blob/947054028b6fbfe58b4ced996f88fc7614ab5cb4/lib/Document/Adapter/Ghostscript.php#L136
https://github.com/pimcore/pimcore/blob/947054028b6fbfe58b4ced996f88fc7614ab5cb4/models/Asset.php#L1118

**Comment by dvesh3** [comment]
Hi @skochTC yes, the [changes ](https://github.com/pimcore/pimcore/pull/15501#issuecomment-1649546546)looks good. I would not touch `ghostscript` & `Asset` as this only get `$this->stream` from asset and there stream is explicitly [closed on destruct ](https://github.com/pimcore/pimcore/blob/9e153e93145a293ee37a91f23a226b84e0d2a229/models/Asset.php#L1686). 

please adapt this PR with your recent suggestions. Thanks!

File: models/Asset/Image/Thumbnail/Processor.php [hunk 1]
```diff
@@ -268,7 +268,9 @@ public static function process(Asset $asset, Config $config, $fileSystemPath = n
                 }
 
                 if (is_resource($fileSystemPath)) {
+                    $fileSystemPathStream = $fileSystemPath;
                     $fileSystemPath = self::getLocalFileFromStream($fileSystemPath);
+                    fclose($fileSystemPathStream);
```
Comment (dvesh3) [inline models/Asset/Image/Thumbnail/Processor.php]:
```suggestion
                    @fclose($fileSystemPathStream);
```
just to be on safe side..

File: models/Asset/Thumbnail/ImageThumbnailTrait.php [hunk 2]
```diff
@@ -403,7 +403,10 @@ public function getLocalFile()
             return null;
         }
 
-        return self::getLocalFileFromStream($stream);
+        $localFile = self::getLocalFileFromStream($stream);
+        fclose($stream);
```
Comment (dvesh3) [inline models/Asset/Thumbnail/ImageThumbnailTrait.php]:
```suggestion
        @fclose($stream);
```

File: models/User.php [hunk 3]
```diff
@@ -566,7 +566,9 @@ public function getImage($width = null, $height = null)
         $storage = Tool\Storage::get('admin');
         if ($storage->fileExists($this->getOriginalImageStoragePath())) {
             if (!$storage->fileExists($this->getThumbnailImageStoragePath())) {
-                $localFile = self::getLocalFileFromStream($storage->readStream($this->getOriginalImageStoragePath()));
+                $originalImageStream = $storage->readStream($this->getOriginalImageStoragePath());
+                $localFile = self::getLocalFileFromStream($originalImageStream);
+                fclose($originalImageStream);
```
Comment (dvesh3) [inline models/User.php]:
```suggestion
                @fclose($originalImageStream);
```

**Review by dvesh3 — APPROVED** [review]

**Comment by dvesh3** [comment]
@skochTC thanks a lot!

**Pull request merged**